### PR TITLE
Propagate Tags to Cloud Formation

### PR DESCRIPTION
**Pull request by pipo02mix**
The goal of this PR is propagating Cluster CR labels (CAPI schema) with the correct format (prefix `aws-tag/`) to Cloud Formation stacks (TCCP, TCPPN, TCPN). 

## Checklist

- [x] Update changelog in CHANGELOG.md.

### CODE REVIEW

**Comment by pipo02mix** [comment]
I was wondering if it is needed to add a cache to avoid each loop calling k8s API to get Cluster CR labels. WDYT?

File: service/internal/cloudtags/spec.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package cloudtags
+
+import "context"
+
+type Interface interface {
+	// AreClusterTagsEquals compares current cluster tags with the stack tags
+	AreClusterTagsEquals(ctx context.Context, ctags map[string]string, stags map[string]string) bool
```
Comment (xh3b4sd) [inline service/internal/cloudtags/spec.go:7]: Suggestions, the name is more idiomatic that way and the interface is simpler, hiding the complexity of having to deal with CRs and Cloud Formation tags. 

```suggestion
	ClusterTagsEqual(ctx context.Context, clusterID string) (bool, error)
```

**Comment by xh3b4sd** [comment]
Note that it is good enough to request review from @giantswarm/team-firecracker-pe in the `aws-operator`. SIG Operator are a lot of people and most of them are not deeply involved in this particular project anymore. See also https://github.com/giantswarm/fmt/blob/master/github/requesting_review.md.

File: service/controller/resource/tcnp/create.go [hunk 2]
```diff
@@ -212,11 +245,17 @@ func (r *Resource) updateStack(ctx context.Context, cr infrastructurev1alpha2.AW
...
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
 		i := &cloudformation.UpdateStackInput{
 			Capabilities: []*string{
 				aws.String(capabilityNamesIAM),
 			},
 			StackName:    aws.String(key.StackNameTCNP(&cr)),
+			Tags:         tags,
```
[the remark this answers is not in the document]
Comment (pipo02mix) [reply service/controller/resource/tcnp/create.go:258]: No, all resources that support tags update this value. but in case of the ASG the machines are only updated when new instances are added to the group

```
 If you modify a tag, the updated version of the tag is added to instances launched in the Auto Scaling group after the change. If you create or modify a tag for an Auto Scaling group, these changes are not made to instances that are already running in the Auto Scaling group.
```
https://docs.aws.amazon.com/autoscaling/ec2/userguide/autoscaling-tagging.html#add-tags

Anyway, I tested multiples times. Also, I check there is no interruption having a pingdom checking service availability of helloworld. So all changed on resources like SG or Subnet does not impact workload traffic

File: service/internal/cloudtags/cloudtags.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+	return true
+}
+
+// GetTagsByCluster the cloud tags from CAPI Cluster CR
+func (ct *CloudTags) GetTagsByCluster(ctx context.Context, clusterID string) (map[string]string, error) {
+	var list apiv1alpha2.ClusterList
+	tags := map[string]string{}
+
+	err := ct.k8sClient.CtrlClient().List(
```
[the remark this answers is not in the document]
Comment (pipo02mix) [reply service/internal/cloudtags/cloudtags.go:70]: yes I was thinking on adding cache, but I did not want to do all in same PR. I will create new PR that points to this branch

File: service/internal/cloudtags/spec.go (hunk 1 above)
Comment (pipo02mix) [reply service/internal/cloudtags/spec.go:7]: At the beginning I wrote it like `AreClusterTagsEquals(ctx context.Context, clusterID, stags map[string]string) bool`
but if we dont inject stack tags then you need to query again AWS API to get those (which we want to avoid). It  would great in terms of hiding complexity but we already call AWS API a lot and I did not want add other extra call (having the tags already on place)

File: service/internal/cloudtags/cloudtags.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+// AreClusterTagsEquals compares current cluster tags with the stack tags
+func (ct *CloudTags) AreClusterTagsEquals(ctx context.Context, ctags map[string]string, stags map[string]string) bool {
+	tagsEqual := reflect.DeepEqual(ctags, stags)
+	if !tagsEqual {
+		ct.logger.LogCtx(ctx,
+			"level", "debug",
+			"message", "detected a change in cloud tags",
+			"reason", fmt.Sprintf("tags changed from %#q to %#q", stags, ctags),
```
[the remark this answers is not in the document]
Comment (pipo02mix) [reply service/internal/cloudtags/cloudtags.go:57]: fair point

File: service/controller/resource/tccp/create.go [hunk 5]
```diff
@@ -107,11 +109,29 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
...
+		if err != nil {
+
+			return microerror.Mask(err)
+		}
+
+		ctags, err := r.cloudtags.GetTagsByCluster(ctx, key.ClusterID(&cr))
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		updateTags := !r.cloudtags.AreClusterTagsEquals(ctx, ctags, stackTags)
```
[the remark this answers is not in the document]
Comment (pipo02mix) [reply service/controller/resource/tccp/create.go:134]: hmm yes I see the point, I did not want to extend changedetection with another dependency, but it makes it more consistent. I will do

File: service/internal/cloudtags/spec.go (hunk 1 above)
Comment (xh3b4sd) [reply service/internal/cloudtags/spec.go:7]: That is what the caching is for, so that it does not matter how often we want to fetch or use the data within a single reconciliation loop.

Comment (pipo02mix) [reply service/internal/cloudtags/spec.go:7]: ok let me take a look to caching code and I will do a PR targeting this branch

**Comment by pipo02mix** [comment]
@xh3b4sd I address all your comments but caching one that comes in a different PR PTAL (pointing to this one to reduce number of changes included)

File: service/controller/resource/tccp/create.go [hunk 6]
```diff
@@ -116,6 +125,11 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 			return microerror.Mask(err)
 		}
 
+		updateTags, err := r.detection.ShouldUpdateTags(ctx, key.ClusterID(&cr), stackTags)
```
Comment (xh3b4sd) [inline service/controller/resource/tccp/create.go:128]: What I would like to see is to use the interfaces we have. The logic should simply be encapsulated in `ShouldUpdate`. Nobody should have to care about something rather complicated like `if key.IsAlreadyCreatedCluster(cr) && updateTags`. All that matters is `if update`.

Comment (xh3b4sd) [reply service/controller/resource/tccp/create.go:128]: Maybe this gets easier when we simply drop all the legacy master handling. All we need should be this at this point. 

```
if update {
	err = r.updateStack(ctx, cr)
	if err != nil {
		return microerror.Mask(err)
	}
}
```

Might be we still need the other code for older versions upgrading to the new version. @calvix can you confirm we can clean this up or do we need to keep the master handling in here?

Comment (xh3b4sd) [reply service/controller/resource/tccp/create.go:128]: Alternatives for the current proposal. Let's maybe think about this for a moment and understand benefits and drawbacks. 

* We could maybe put the tags directly into the Cloud Formation stacks, if that is possible. 
* We could have a separate handler for the tag propagation.

File: service/internal/cloudtags/spec.go [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package cloudtags
+
+import "context"
+
+type Interface interface {
+	// ClusterLabelsNotEqual compares current cluster labels with the stack tags passed
+	ClusterLabelsNotEqual(ctx context.Context, clusterID string, stags map[string]string) (bool, error)
```
Comment (xh3b4sd) [inline service/internal/cloudtags/spec.go:7]: * It is rather unfortunate to negate the boolean with a function name. Something should rather return true so that you can simply do `!true` in the code if you have to. People are used to that and understand it more easily. Having `Not` in the same forces me to very carefully read and think. 
* The interface should be clean. The tags should not be given, but managed internally. 

```suggestion
	ClusterLabelsEqual(ctx context.Context, clusterID string) (bool, error)
```

Comment (pipo02mix) [reply service/internal/cloudtags/spec.go:7]: Indeed I change it because I thought `!` was easier to miss than a negative expression. But I can revert it easily :)

**Comment by kopiczko** [comment]
This is unrelated but do we need SIG operator review here?

**Comment by xh3b4sd** [comment]
> This is unrelated but do we need SIG operator review here?

See https://github.com/giantswarm/aws-operator/pull/2547#issuecomment-655306945.

**Comment by xh3b4sd** [comment]
Can we merge or close this?

**Comment by pipo02mix** [comment]
I did not have time to work on this and then holidays. I have the picture how to do it, but still, need to invest more time as the dependency injection is quite tricky at the moment. Also, resolve conflicts from the latest changes 😅

File: service/controller/resource/tcnp/create.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (calvix) [reply service/controller/resource/tcnp/create.go:250]: I think there could be corner cases that this would trigger a node roll, if there is already some changes on CR  that are not applied yet but would cause node roll than this harmless Tag update would trigger the changes as well.

File: service/controller/resource/tccp/create.go [hunk 8]
```diff
@@ -121,6 +121,11 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 			if err != nil {
 				return microerror.Mask(err)
 			}
+		} else if key.IsAlreadyCreatedCluster(cr) {
```
Comment (calvix) [inline service/controller/resource/tccp/create.go:124]: It will be unpredictable if we trigger a possible CF update every reconciliation loop, as mentioned below there could be  changes that are not applied yet and it would immediately trigger a change


Also  in the case of many cluster, we can easily hit the AWS API call limit if we try   an update every reconciliation loop

**Comment by calvix** [comment]
I personally don't like having random or accident updates that could be triggered by someone messing with tags/label.

There could be a scenario where someone will remove a tag from AWS resource and it will trigger and node roll. IMHO that's not OK.


I would leave the update logic the same as before and let tags update only when there is an actual update. It might feel slow but how much often customers ned to change tags on the clusters? Do they really need to do it every day or is it that they set it once and it stays the same?

File: service/controller/resource/tcnp/create.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pipo02mix) [reply service/controller/resource/tcnp/create.go:250]: But if there is a change in node pool CR, it should not be the case? I mean, when you change the CR spec, operator should trigger the changes, right?

File: service/controller/resource/tccp/create.go (hunk 8 above)
Comment (pipo02mix) [reply service/controller/resource/tccp/create.go:124]: but it is the reason why I implemented the AWS cache, to save API requests

**Comment by pipo02mix** [comment]
> I would leave the update logic the same as before and let tags update only when there is an actual update. It might feel slow but how much often customers need to change tags on the clusters? Do they really need to do it every day or is it that they set it once and it stays the same?

Ok I like your idea, let's start small. We only set tags at creation time and when CRs trigger updates to the stacks! I change my code, actually it makes it easier :)

File: service/controller/resource/tccpazs/create_test.go [hunk 9]
```diff
@@ -235,8 +235,8 @@ func Test_EnsureCreated_AZ_Spec(t *testing.T) {
 					t.Fatal(err)
 				}
 
-				for _, md := range tc.machineDeployments {
-					err := r.k8sClient.CtrlClient().Create(ctx, &md)
+				for i := range tc.machineDeployments {
+					err := r.k8sClient.CtrlClient().Create(ctx, &tc.machineDeployments[i])
```
Comment (pipo02mix) [inline service/controller/resource/tccpazs/create_test.go:239]: I fix a `go vet` issue here

**Comment by pipo02mix** [comment]
I have followed @calvix advise, which has simplified the code a lot. Now tags are only considered when the stack is going to be created or updated. 

Also now there is no cache due to this operation (create/update stack) are very rare.

PTAL it would be great to include in 14.x.x

File: service/internal/changedetection/tcnp.go [hunk 10]
```diff
@@ -44,8 +44,8 @@ func NewTCNP(config TCNPConfig) (*TCNP, error) {
 	}
 
 	t := &TCNP{
-		event:    config.Event,
 		logger:   config.Logger,
+		event:    config.Event,
```
Comment (calvix) [inline service/internal/changedetection/tcnp.go:48]: what happened here :trollface: I think  order by alphabet broke here :trollface:

File: service/internal/cloudtags/cloudtags.go [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiv1alpha2 "sigs.k8s.io/cluster-api/api/v1alpha2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/aws-operator/pkg/label"
+)
+
+const keyCloudPrefix = "aws-tag/"
```
Comment (calvix) [inline service/internal/cloudtags/cloudtags.go:17]: is this something that is from upstream or is this something we invented? was this consulted with sig-ux and sig-product? 

Just wanted to make sure we don't add some prefix that someone won't like a few weeks later due to bad UX :)

**Review by calvix — APPROVED** [review]
code looks good, LGTM if tested, just a small comment about the aws-tag prefix

**Comment by calvix** [comment]
also, changelog entry would be nice

File: pkg/project/project.go [hunk 12]
```diff
@@ -5,7 +5,7 @@ var (
 	gitSHA             = "n/a"
 	name        string = "aws-operator"
 	source      string = "https://github.com/giantswarm/aws-operator"
-	version            = "10.0.1-dev"
+	version            = "10.1.1"
```
Comment (xh3b4sd) [inline pkg/project/project.go:8]: That looks wrong. Only the Github Action for releases should change the version like this. If you want to test you can customize the suffix to your liking.

File: service/internal/cloudtags/cloudtags.go (hunk 11 above)
Comment (pipo02mix) [reply service/internal/cloudtags/cloudtags.go:17]: yeah good point. In upstream there is no such a tag, for example, AWS cloud provider uses `service.beta.kubernetes.io/aws-load-balancer-additional-resource-tags` label for tagging LBs. In [cluster API I did not find anything](https://cluster-api-aws.sigs.k8s.io/topics/consuming-existing-aws-infrastructure.html?highlight=tag#tagging-aws-resources)

**Review by xh3b4sd — COMMENTED** [review]
FMI where does `aws-tag/` come from? Is that an upstream convention? Usually there is a domain structure defined like `tag.aws.io/`.

File: service/internal/cloudtags/cloudtags.go (hunk 11 above)
Comment (pipo02mix) [reply service/internal/cloudtags/cloudtags.go:17]: @giantswarm/sig-ux PTAL to the naming. `aws-tag/` would be the prefix for propagating tags to cloud provider resource (example `aws-tag/my-tag-key: my-value`). Anything more appropriate?

Comment (pipo02mix) [reply service/internal/cloudtags/cloudtags.go:17]: > FMI where does aws-tag/ come from? Is that an upstream convention? Usually, there is a domain structure defined like tag.aws.io/.

@xh3b4sd let's continue here, I am fine with aligning to domain format `tag.aws.io/`

File: CHANGELOG.md [hunk 13]
```diff
@@ -16,6 +16,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   roles.
 - Update `k8scloudconfig` version to `v10.0.0` to include change for Kubernetes 1.19.
 - Allow configuration of `MINIMUM_IP_TARGET` and `WARM_IP_TARGET` for AWS CNI via annotations on `AWSCluster`
+- Add support for tagging AWS resources, managed by the operator, based on the custom resource labels.
```
Comment (calvix) [inline CHANGELOG.md:19]: 10.0.0 is already released you need to add the  changelog into the `Unreleased` part

Comment (pipo02mix) [reply CHANGELOG.md]: You I right, I miss read it

File: service/internal/cloudtags/cloudtags.go (hunk 11 above)
Comment (pipo02mix) [reply service/internal/cloudtags/cloudtags.go:17]: We discussed in SIG UX and we decided to go with `tag.provider.giantswarm.io/`

**Comment by pipo02mix** [comment]
@xh3b4sd PTAL

File: service/internal/cloudtags/cloudtags.go [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+// IsCloudTagKey check is a tag with proper prefix
+func isCloudTagKey(tagKey string) bool {
+	return strings.HasPrefix(tagKey, keyCloudPrefix)
+}
+
+// TrimCloudTagKey check is a tag with proper prefix
+func trimCloudTagKey(tagKey string) string {
+	return strings.Replace(tagKey, keyCloudPrefix, "", 1)
```
Comment (xh3b4sd) [inline service/internal/cloudtags/cloudtags.go:96]: IMO it would be safer to use https://golang.org/pkg/strings/#TrimPrefix. Reason is that if we ever have `foo.tag.provider.giantswarm.io/` you end up with `foo.` which would likely cause some nasty issues.

Comment (pipo02mix) [reply service/internal/cloudtags/cloudtags.go]: done, thx for the advise!

Comment (xh3b4sd) [reply service/internal/cloudtags/cloudtags.go]: You could add a test case for that if you are fancy. 😅

**Review by xh3b4sd — APPROVED** [review]

File: service/internal/cloudtags/cloudtags.go (hunk 14 above)
Comment (pipo02mix) [reply service/internal/cloudtags/cloudtags.go]: There is some test here https://github.com/giantswarm/aws-operator/pull/2547/files#diff-863ef0fba22bf708bfe6231bffdd376d17ac26ed503b7a6defff241f0ecb27bfR35 that check it returns expected values :)

Comment (xh3b4sd) [reply service/internal/cloudtags/cloudtags.go]: These tests were written for `strings.Replace` to work. There is no test covering the edge case of `strings.TrimPrefix` as discussed above.
